File: leadera/src/main/java/com/leadera/leadera/service/TareaService.java
package com.leadera.leadera.service;

import com.leadera.leadera.dto.AsistenteResumenDTO;
import com.leadera.leadera.dto.CrearTareaRequest;
import com.leadera.leadera.dto.TareaDTO;
import com.leadera.leadera.entity.Agente;
import com.leadera.leadera.entity.Lead;
import com.leadera.leadera.entity.Tarea;
import com.leadera.leadera.enums.RolAgente;
import com.leadera.leadera.exception.ResourceNotFoundException;
import com.leadera.leadera.exception.UnauthorizedActionException;
import com.leadera.leadera.repository.AgenteRepository;
import com.leadera.leadera.repository.InteraccionRepository;
import com.leadera.leadera.repository.LeadRepository;
import com.leadera.leadera.repository.TareaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;

@Service
public class TareaService {

    private final TareaRepository tareaRepository;
    private final InteraccionRepository interaccionRepository;
    private final AgenteRepository agenteRepository;
    private final LeadRepository leadRepository;
    private final AgenteContextResolver agenteContextResolver;
    private final ZoneId zonaHoraria;

    public TareaService(TareaRepository tareaRepository,
                        InteraccionRepository interaccionRepository,
                        AgenteRepository agenteRepository,
                        LeadRepository leadRepository,
                        AgenteContextResolver agenteContextResolver,
                        ZoneId zonaHoraria) {
        this.tareaRepository = tareaRepository;
        this.interaccionRepository = interaccionRepository;
        this.agenteRepository = agenteRepository;
        this.leadRepository = leadRepository;
        this.agenteContextResolver = agenteContextResolver;
        this.zonaHoraria = zonaHoraria;
    }

    // 3a — El AGENTE/DUENO crea una tarea para uno de SUS asistentes.
    // Un ASISTENTE también puede crear, pero solo autoasignándosela.
    @Transactional
    public TareaDTO crearTarea(CrearTareaRequest request, String emailCreador) {
        Agente actor = agenteContextResolver.resolverActor(emailCreador);

        // Un asistente solo se autoasigna tareas (no puede crearle a nadie más).
        if (actor.getRol() == RolAgente.ASISTENTE
                && !actor.getId().equals(request.getAsignadoAId())) {
            throw new UnauthorizedActionException("Un asistente solo puede asignarse tareas a sí mismo.");
        }

        Agente asignadoA = agenteRepository.findById(request.getAsignadoAId())
                .orElseThrow(() -> new ResourceNotFoundException("Agente asignado no encontrado"));

        // Cross-tenant: el asignado debe ser de la misma inmobiliaria que el creador.
        // (Para el asistente autoasignándose es trivialmente cierto.)
        if (!asignadoA.getInmobiliaria().getId().equals(actor.getInmobiliaria().getId())) {
            throw new UnauthorizedActionException("No podés asignar tareas a agentes de otra inmobiliaria.");
        }

        // Jerarquía (solo para AGENTE/DUENO): se asigna únicamente a los propios
        // asistentes (supervisor == actor). No aplica a la autoasignación del
        // asistente: su supervisor es su agente, no él mismo.
        if (actor.getRol() != RolAgente.ASISTENTE
                && (asignadoA.getSupervisor() == null
                        || !asignadoA.getSupervisor().getId().equals(actor.getId()))) {
            throw new UnauthorizedActionException("Solo podés asignar tareas a tus propios asistentes.");
        }

        Tarea tarea = new Tarea();
        tarea.setTitulo(request.getTitulo());
        tarea.setAsignadoA(asignadoA);
        tarea.setCreadoPor(actor);
        tarea.setObjetivo(request.getObjetivo());
        tarea.setFechaObjetivo(request.getFechaObjetivo());
        tarea.setInmobiliaria(actor.getInmobiliaria());
        tarea.setFechaCreacion(LocalDateTime.now(zonaHoraria));

        if (request.getLeadId() != null) {
            Lead lead = leadRepository.findById(request.getLeadId())
                    .orElseThrow(() -> new ResourceNotFoundException("Lead no encontrado"));
            // El lead debe pertenecer al PROPIETARIO EFECTIVO del creador (regla de
            // Fase 1): para AGENTE/DUENO es él mismo; para un asistente que se
            // autoasigna, los leads que trabaja son los de su supervisor. Cierra el
            // cruce de inmobiliaria de forma implícita.
            Agente propietario = agenteContextResolver.resolverPropietario(actor);
            if (!lead.getAgente().getId().equals(propietario.getId())) {
                throw new UnauthorizedActionException("No podés asociar la tarea a un lead que no es tuyo.");
            }
            tarea.setLead(lead);
        }

        Tarea guardada = tareaRepository.save(tarea);
        return toDTO(guardada, calcularProgreso(guardada, asignadoA.getId()));
    }

    // 3b — El asistente (o cualquier actor) ve SUS tareas de hoy; las cuotas se
    // auto-completan si alcanzaron el objetivo.
    @Transactional
    public List<TareaDTO> obtenerTareasDelDia(String emailActor) {
        Agente actor = agenteContextResolver.resolverActor(emailActor);

        LocalDateTime inicioHoy = LocalDateTime.now(zonaHoraria).toLocalDate().atStartOfDay();
        LocalDateTime finHoyIncl = inicioHoy.plusDays(1).minusNanos(1);

        List<Tarea> tareas = tareaRepository.findByAsignadoAIdAndFechaObjetivoBetween(
                actor.getId(), inicioHoy, finHoyIncl);

        List<TareaDTO> salida = new ArrayList<>(tareas.size());
        for (Tarea tarea : tareas) {
            int progreso = calcularProgreso(tarea, actor.getId());

            // Auto-completado de cuotas: si llegó al objetivo y aún no estaba marcada.
            if (tarea.getObjetivo() != null && !tarea.isCompletada() && progreso >= tarea.getObjetivo()) {
                tarea.setCompletada(true);
                tarea.setFechaCompletada(LocalDateTime.now(zonaHoraria));
                tareaRepository.save(tarea);
            }

            salida.add(toDTO(tarea, progreso));
        }
        return salida;
    }

    // 3c — El asistente marca a mano una tarea DISCRETA propia.
    @Transactional
    public TareaDTO completarTarea(Long tareaId, String emailActor) {
        Agente actor = agenteContextResolver.resolverActor(emailActor);

        // Cross-tenant: la tarea debe ser de la inmobiliaria del actor.
        Tarea tarea = tareaRepository.findByIdAndInmobiliariaId(tareaId, actor.getInmobiliaria().getId())
                .orElseThrow(() -> new ResourceNotFoundException("Tarea no encontrada"));

        // Solo el asignado puede completarla.
        if (!tarea.getAsignadoA().getId().equals(actor.getId())) {
            throw new UnauthorizedActionException("No tenés permiso para completar esta tarea.");
        }

        // Las cuotas se completan solas; no se marcan a mano.
        if (tarea.getObjetivo() != null) {
            throw new UnauthorizedActionException("Esta tarea se completa automáticamente.");
        }

        tarea.setCompletada(true);
        tarea.setFechaCompletada(LocalDateTime.now(zonaHoraria));
        Tarea guardada = tareaRepository.save(tarea);

        return toDTO(guardada, 0);
    }

    // 3d — El supervisor ve las tareas que le creó a un asistente suyo.
    @Transactional(readOnly = true)
    public List<TareaDTO> obtenerTareasDeAsistente(Long asistenteId, String emailSupervisor) {
        Agente supervisor = agenteContextResolver.resolverActor(emailSupervisor);

        Agente asistente = agenteRepository.findById(asistenteId)
                .orElseThrow(() -> new ResourceNotFoundException("Asistente no encontrado"));

        // Cross-tenant + jerarquía: mismo tenant y supervisor == el que pide.
        boolean mismaInmobiliaria = asistente.getInmobiliaria().getId()
                .equals(supervisor.getInmobiliaria().getId());
        boolean esSuAsistente = asistente.getSupervisor() != null
                && asistente.getSupervisor().getId().equals(supervisor.getId());
        if (!mismaInmobiliaria || !esSuAsistente) {
            throw new UnauthorizedActionException("Este asistente no está bajo tu supervisión.");
        }

        List<Tarea> tareas = tareaRepository.findByCreadoPorIdAndAsignadoAId(supervisor.getId(), asistenteId);

        List<TareaDTO> salida = new ArrayList<>(tareas.size());
        for (Tarea tarea : tareas) {
            // Solo se calcula progreso; acá NO se auto-completa (vista de lectura del supervisor).
            salida.add(toDTO(tarea, calcularProgreso(tarea, asistenteId)));
        }
        return salida;
    }

    // Asistentes del actor (para el selector del frontend). Lista vacía si no tiene.
    @Transactional(readOnly = true)
    public List<AsistenteResumenDTO> obtenerMisAsistentes(String emailActor) {
        Agente actor = agenteContextResolver.resolverActor(emailActor);
        return agenteRepository.findBySupervisorId(actor.getId()).stream()
                .map(a -> new AsistenteResumenDTO(a.getId(), a.getNombre(), a.getApellido()))
                .toList();
    }

    // ---- helpers ----

    // Progreso de una cuota: leads distintos contactados HOY por el asistente (autor).
    // Para tareas discretas (objetivo == null) el progreso no aplica: devuelve 0.
    private int calcularProgreso(Tarea tarea, Long autorId) {
        if (tarea.getObjetivo() == null) {
            return 0;
        }
        LocalDateTime inicioHoy = LocalDateTime.now(zonaHoraria).toLocalDate().atStartOfDay();
        LocalDateTime finHoy = inicioHoy.plusDays(1);
        return (int) interaccionRepository.countLeadsDistinctPorAutorEnRango(autorId, inicioHoy, finHoy);
    }

    private TareaDTO toDTO(Tarea tarea, int progreso) {
        Agente asignadoA = tarea.getAsignadoA();
        String nombre = ((asignadoA.getNombre() != null ? asignadoA.getNombre() : "") + " "
                + (asignadoA.getApellido() != null ? asignadoA.getApellido() : "")).trim();
        return new TareaDTO(
                tarea.getId(),
                tarea.getTitulo(),
                asignadoA.getId(),
                nombre,
                tarea.getCreadoPor().getId(),
                tarea.getLead() != null ? tarea.getLead().getId() : null,
                tarea.getObjetivo(),
                progreso,
                tarea.isCompletada(),
                tarea.getFechaCompletada(),
                tarea.getFechaObjetivo()
        );
    }
}

File: leadera-front/src/app/core/services/tarea-service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

// Espejo del TareaDTO del backend (Fase 2).
export interface TareaDTO {
  id: number;
  titulo: string;
  asignadoAId: number;
  asignadoANombre: string;
  creadoPorId: number;
  leadId: number | null;
  // null = tarea discreta (manual); con número = cuota automática.
  objetivo: number | null;
  // Leads distintos contactados hoy por el asistente (0 si la tarea es discreta).
  progreso: number;
  completada: boolean;
  fechaCompletada: string | null;
  fechaObjetivo: string;
}

// Espejo del AsistenteResumenDTO del backend.
export interface AsistenteResumen {
  id: number;
  nombre: string;
  apellido: string;
}

// Espejo del AsistenteStatsDTO del backend: fila de "Mi equipo" del agente.
// leadsActivos son los del supervisor (el asistente trabaja sobre esos leads).
export interface AsistenteStats {
  id: number;
  nombre: string;
  apellido: string;
  email: string;
  activo: boolean;
  leadsActivos: number;
  tareasCompletadasHoy: number;
}

// Espejo de CrearAgenteEquipoRequest (mismo DTO que usa el alta de agentes del dueño).
export interface CrearAsistenteRequest {
  nombre: string;
  apellido: string;
  email: string;
  passwordTemporal: string;
}

// Respuesta del alta de asistente: el backend reutiliza AgenteEquipoDTO
// (trae rol en lugar de tareasCompletadasHoy).
export interface AsistenteCreado {
  id: number;
  nombre: string;
  apellido: string;
  email: string;
  rol: string;
  activo: boolean;
  leadsActivos: number;
}

// Espejo de CrearTareaRequest del backend.
export interface CrearTareaRequest {
  titulo: string;
  asignadoAId: number;
  leadId?: number | null;
  // null/ausente = tarea discreta; con número = cuota automática.
  objetivo?: number | null;
  // LocalDateTime ISO (ej: "2026-06-11T23:59:59").
  fechaObjetivo: string;
}

@Injectable({
  providedIn: 'root',
})
export class TareaService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/tareas`;
  private asistentesUrl = `${environment.apiUrl}/agente/asistentes`;

  // Tareas que el supervisor le creó a uno de sus asistentes.
  obtenerTareasDeAsistente(asistenteId: number): Observable<TareaDTO[]> {
    return this.http.get<TareaDTO[]>(`${this.apiUrl}/asistente/${asistenteId}`);
  }

  // Asistentes del actor autenticado (lista vacía si no tiene).
  obtenerMisAsistentes(): Observable<AsistenteResumen[]> {
    return this.http.get<AsistenteResumen[]>(`${this.apiUrl}/mis-asistentes`);
  }

  // Marca a mano una tarea discreta (acción del asistente, no del supervisor).
  completarTarea(tareaId: number): Observable<TareaDTO> {
    return this.http.patch<TareaDTO>(`${this.apiUrl}/${tareaId}/completar`, {});
  }

  // Tareas de HOY del actor autenticado (vista del asistente en /gestion-del-dia).
  obtenerTareasDelDia(): Observable<TareaDTO[]> {
    return this.http.get<TareaDTO[]>(`${this.apiUrl}/hoy`);
  }

  // Crea una tarea (AGENTE/DUENO para su asistente; un asistente solo se autoasigna).
  crearTarea(request: CrearTareaRequest): Observable<TareaDTO> {
    return this.http.post<TareaDTO>(this.apiUrl, request);
  }

  // ---- Gestión de asistentes (vista "Mi equipo" del agente) ----

  // Asistentes del actor con métricas del día (reemplaza a obtenerMisAsistentes
  // en la página; el sidebar sigue usando la versión liviana).
  listarAsistentes(): Observable<AsistenteStats[]> {
    return this.http.get<AsistenteStats[]>(this.asistentesUrl);
  }

  crearAsistente(request: CrearAsistenteRequest): Observable<AsistenteCreado> {
    return this.http.post<AsistenteCreado>(this.asistentesUrl, request);
  }

  cambiarActivoAsistente(asistenteId: number, activo: boolean): Observable<AsistenteStats> {
    return this.http.patch<AsistenteStats>(`${this.asistentesUrl}/${asistenteId}/activo`, { activo });
  }
}
